### PR TITLE
Fix typing cqr

### PR DESCRIPTION
**Pull request by LacombeLouis**
# Description
There was an issue with the typing from the fit and predict. In order to fix it, we will look at the problem from a bigger picture, the goal is to solve this issue while also making `MapieQuantileRegressor` more like `MapieRegressor`. This fixes issue #176.

We will allow the `fit` method to only have `X` and `y` as required variables, meaning that if `X_calib` and `y_calib` are `None`(as set by default), then the `train_test_split` will occur inside the fit function.

## Type of change

Please remove options that are irrelevant.

- Bug fix (non-breaking change which fixes an issue)

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

File: mapie/quantile_regression.py [hunk 1]
```diff
@@ -316,13 +321,14 @@ def _check_cv(
...
 
-    def fit(  # type: ignore
+    def fit(
         self,
         X: ArrayLike,
         y: ArrayLike,
-        X_calib: ArrayLike,
-        y_calib: ArrayLike,
         sample_weight: Optional[ArrayLike] = None,
+        X_calib: Optional[ArrayLike] = None,
```
Comment (vtaquet) [inline mapie/quantile_regression.py:329]: Rearrange the list of arguments in the docstring accordingly. 
Add `calib_size` in docstring.

File: mapie/quantile_regression.py [hunk 2]
```diff
@@ -363,16 +369,41 @@ def fit(  # type: ignore
...
+                )
+            else:
+                X_train, X_calib, y_train, y_calib = train_test_split(
+                    X,
+                    y,
+                    test_size=calib_size
+                )
+                sample_weight_train = sample_weight
+        else:
+            X_train, y_train, sample_weight_train = X, y, sample_weight
```
Comment (vtaquet) [inline mapie/quantile_regression.py:395]: I would move this block in a dedicated hidden method.

File: mapie/quantile_regression.py [hunk 3]
```diff
@@ -363,16 +369,41 @@ def fit(  # type: ignore
         alpha = self._check_alpha(self.alpha)
         self.cv = self._check_cv(cast(str, self.cv))
         X, y = indexable(X, y)
+        if X_calib is None or y_calib is None:
+            if sample_weight is not None:
```
Comment (vtaquet) [inline mapie/quantile_regression.py:373]: I would start with `if sample_weight is None`

File: mapie/quantile_regression.py [hunk 4]
```diff
@@ -363,16 +369,41 @@ def fit(  # type: ignore
...
+                    X,
+                    y,
+                    test_size=calib_size
+                )
+                sample_weight_train = sample_weight
+        else:
+            X_train, y_train, sample_weight_train = X, y, sample_weight
+        X_train, y_train = indexable(X_train, y_train)
         X_calib, y_calib = indexable(X_calib, y_calib)
         y, y_calib = _check_y(y), _check_y(y_calib)
```
Comment (vtaquet) [inline mapie/quantile_regression.py:398]: `y_train` instead of `y` ?

File: mapie/quantile_regression.py [hunk 5]
```diff
@@ -453,6 +486,11 @@ def predict(  # type: ignore
             - [:, 1, :]: Upper bound of the prediction interval.
         """
         check_is_fitted(self, self.fit_attributes)
+        if alpha is not None:
```
Comment (vtaquet) [inline mapie/quantile_regression.py:489]: Shall we raise a warning instead by saying that the `alpha` defined in `predict` won't be used for PI estimates and then fix this `alpha` to None ?

File: mapie/quantile_regression.py [hunk 6]
```diff
@@ -453,6 +486,11 @@ def predict(  # type: ignore
...
+            raise ValueError(
+                "Alpha should not be specified in the prediction method"
+                + "with CQR"
```
Comment (vtaquet) [inline mapie/quantile_regression.py:492]: Don't use the acronym in error/warning messages as the message needs to be self-sufficient.

File: mapie/tests/test_quantile_regression.py [hunk 7]
```diff
@@ -320,18 +354,29 @@ def test_wrong_alphas_types(alphas: float) -> None:
         match=r".*Invalid alpha. Allowed values are float.*",
     ):
         mapie_reg = MapieQuantileRegressor(alpha=alphas)
-        mapie_reg.fit(X_train, y_train, X_calib, y_calib)
+        mapie_reg.fit(X_train, y_train, X_calib=X_calib, y_calib=y_calib)
+
+
+def test_alpha_in_predict() -> None:
+    """Checking for alpha in predict"""
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:361]: Checking for alpha in predict raises a warning/error.

File: mapie/tests/test_quantile_regression.py [hunk 8]
```diff
@@ -354,14 +399,41 @@ def test_invalid_cv(cv: Any) -> None:
...
-    mapie.fit(X_train_toy, y_train_toy, X_calib_toy, y_calib_toy)
+    mapie.fit(
+        X_train_toy,
+        y_train_toy,
+        X_calib=X_calib_toy,
+        y_calib=y_calib_toy
+    )
+
+
+def test_X_calib_isNone() -> None:
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:422]: No capital letters in functions.

File: mapie/tests/test_quantile_regression.py [hunk 9]
```diff
@@ -354,14 +399,41 @@ def test_invalid_cv(cv: Any) -> None:
...
+    """Test that the fit method works when X_calib or y_calib is None."""
+    mapie = MapieQuantileRegressor()
+    mapie.fit(X, y, calib_size=0.5)
+    mapie.predict(X)
+
+
+def test_X_calib_isNone_with_sample_weight() -> None:
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:429]: No capital letters in functions.

File: mapie/tests/test_quantile_regression.py [hunk 10]
```diff
@@ -354,14 +399,41 @@ def test_invalid_cv(cv: Any) -> None:
...
+    """
+    Test that the fit method works with calib dataset defined is None
+    with sample weights.
+    """
+    mapie = MapieQuantileRegressor()
+    mapie.fit(X, y, calib_size=0.5, sample_weight=np.ones(X.shape[0]))
+    mapie.predict(X)
```
Comment (vtaquet) [inline mapie/tests/test_quantile_regression.py:436]: Is it possible to test the predictions between 1. a train_test_split done by the user (with X_train_toy, X_calib_toy, y_train_toy, y_calib_toy), 2. a train_test_split performed within MAPIE ?

File: mapie/quantile_regression.py [hunk 11]
```diff
@@ -363,16 +369,41 @@ def fit(  # type: ignore
...
+        if X_calib is None or y_calib is None:
+            if sample_weight is not None:
+                (
+                    X_train,
+                    X_calib,
+                    y_train,
+                    y_calib,
+                    sample_weight_train,
+                    _,
+                ) = train_test_split(
```
Comment (vtaquet) [inline mapie/quantile_regression.py:381]: Be careful, `random_state`, `shuffle`, and `stratify` are not defined and should be added as additional arguments in the fit method.

File: mapie/quantile_regression.py (hunk 4 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:398]: indeed, done.

File: mapie/quantile_regression.py (hunk 3 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:373]: done

File: mapie/quantile_regression.py (hunk 1 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:329]: yes, done

File: mapie/quantile_regression.py (hunk 6 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:492]: done

File: mapie/tests/test_quantile_regression.py (hunk 8 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:422]: done

File: mapie/tests/test_quantile_regression.py (hunk 9 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:429]: done

File: mapie/quantile_regression.py (hunk 11 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:381]: done

File: mapie/tests/test_quantile_regression.py (hunk 10 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:436]: Yes, with implementation of `random_state`, `shuffle` and `stratify` that is now possible. Done.

File: mapie/quantile_regression.py (hunk 5 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:489]: I think leaving a warning is indeed fine, otherwise, I would not bother to change the value necessarily since we get the `alpha` later from `self.alpha`.

File: mapie/tests/test_quantile_regression.py (hunk 7 above)
Comment (LacombeLouis) [reply mapie/tests/test_quantile_regression.py:361]: done

File: mapie/quantile_regression.py (hunk 2 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:395]: done

File: mapie/quantile_regression.py [hunk 12]
```diff
@@ -316,13 +323,82 @@ def _check_cv(
...
+        y_calib: Optional[ArrayLike] = None,
+        calib_size: Optional[float] = 0.3,
+        random_state: Optional[Union[int, np.random.RandomState, None]] = None,
+        shuffle: Optional[bool] = True,
+        stratify: Optional[ArrayLike] = None,
+    ) -> Tuple[
+        ArrayLike, ArrayLike, ArrayLike, ArrayLike, Optional[ArrayLike]
+    ]:
+        """
+        HPUT SOMETHING
```
Comment (vtaquet) [inline mapie/quantile_regression.py:341]: Summary to add.

File: mapie/quantile_regression.py [hunk 13]
```diff
@@ -316,13 +323,82 @@ def _check_cv(
...
+
+        Parameters
+        ----------
+        Same definition of parameters as for the ``fit`` method.
+
+        Returns
+        -------
+        Tuple[ArrayLike, ArrayLike, ArrayLike, ArrayLike, ArrayLike]
+            - [0]: ArrayLike, X_train
```
Comment (vtaquet) [inline mapie/quantile_regression.py:350]: Add the shape ?

File: mapie/utils.py [hunk 14]
```diff
@@ -514,3 +514,52 @@ def check_conformity_score(
...
         )
+
+
+def check_defined_variables_predict_cqr(
+    ensemble: bool,
+    alpha: Union[float, Iterable[float], None],
+) -> None:
+    """
+    Check that the parameters defined for the predict method
+    of the conformalized quantile regression are correct.
```
Comment (vtaquet) [inline mapie/utils.py:525]: of ``MapieQuantileRegressor``

File: mapie/utils.py [hunk 15]
```diff
@@ -514,3 +514,52 @@ def check_conformity_score(
             "Invalid conformity_score argument.\n"
             "Must be None or a ConformityScore instance."
         )
+
+
+def check_defined_variables_predict_cqr(
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/utils.py:519]: The issue is that we would have to import `warnings` into `MapieQuantileRegressor` which isn't added in any other method? So then, we would have to set to be automatically downloaded.

File: mapie/quantile_regression.py (hunk 13 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:350]: done
e.g. 
for `X_train`: `[0]: ArrayLike of shape (n_samples_*(1-calib_size), n_features)`
or 
for `X_calib`: `[2]: ArrayLike of shape (n_samples_*calib_size, n_features)`

Is that the correct method?

File: mapie/quantile_regression.py (hunk 12 above)
Comment (LacombeLouis) [reply mapie/quantile_regression.py:341]: done

File: mapie/quantile_regression.py (hunk 13 above)
Comment (vtaquet) [reply mapie/quantile_regression.py:350]: Yes perfect !

File: mapie/utils.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (vtaquet) [reply mapie/utils.py:519]: Yes you're right, let's leave it here then.
